e2e: mark "etcdserver: request timed out" errors as retryable

nixpanic opened the pull request:
There are regular CI failures where etcdserver times out. These errors
seem not to get caught by any of the existing error comparing. Matching
the error by string should prevent temporary etcdserver issues now too.

Updates: #2218
Closes: #1969

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

- `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)
- `/retest all`: run this in case the CentOS CI failed to start/report any test
  progress or results

</details>

nixpanic committed:
e2e: mark "etcdserver: request timed out" errors as retryable

There are regular CI failures where etcdserver times out. These errors
seem not to get caught by any of the existing error comparing. Matching
the error by string should prevent temporary etcdserver issues now too.

Updates: #2218
Closes: #1969
Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic commented:
#2218 was updated to remove the string comparing as it was assumed that the error comparing was sufficient. Unfortunately it was not.

nixpanic committed:
e2e: only call error check functions when err != nil

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic commented:
The 2nd commit should address the following panic:
```
    k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
    	/go/src/github.com/ceph/ceph-csi/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:55 +0x109
    panic(0x1cec6a0, 0x3176e00)
    	/usr/local/go/src/runtime/panic.go:965 +0x1b9
    github.com/ceph/ceph-csi/e2e.isRetryableAPIError(0x0, 0x0, 0xc00018c018)
    	/go/src/github.com/ceph/ceph-csi/e2e/errors.go:39 +0x127
    github.com/ceph/ceph-csi/e2e.createPVCAndvalidatePV.func1(0xc000554674, 0x2, 0x2)
    	/go/src/github.com/ceph/ceph-csi/e2e/pvc.go:66 +0x47c
    k8s.io/apimachinery/pkg/util/wait.runConditionWithCrashProtection(0xc000554858, 0xc000554600, 0x0, 0x0)
    	/go/src/github.com/ceph/ceph-csi/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:211 +0x69
    k8s.io/apimachinery/pkg/util/wait.WaitFor(0xc000046e10, 0xc000554858, 0xc0005e65a0, 0x0, 0x0)
    	/go/src/github.com/ceph/ceph-csi/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:541 +0x128
    k8s.io/apimachinery/pkg/util/wait.pollInternal(0xc000046e10, 0xc000554858, 0x0, 0x0)
    	/go/src/github.com/ceph/ceph-csi/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:427 +0x87
    k8s.io/apimachinery/pkg/util/wait.pollImmediateInternal(0xc000046e10, 0xc000554858, 0xc000046e10, 0x1)
    	/go/src/github.com/ceph/ceph-csi/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:452 +0x74
    k8s.io/apimachinery/pkg/util/wait.PollImmediate(0x77359400, 0x8bb2c97000, 0xc000554858, 0x25, 0xc000554838)
    	/go/src/github.com/ceph/ceph-csi/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:441 +0x4d
    github.com/ceph/ceph-csi/e2e.createPVCAndvalidatePV(0x22e6588, 0xc000b546e0, 0xc0003f7a40, 0xa, 0x0, 0xc000d8f000)
    	/go/src/github.com/ceph/ceph-csi/e2e/pvc.go:44 +0x3f4
    github.com/ceph/ceph-csi/e2e.createPVCAndApp(0x0, 0x0, 0xc000385e40, 0xc0003f76c0, 0xc000d8f000, 0xa, 0x3e, 0x518a79)
    	/go/src/github.com/ceph/ceph-csi/e2e/utils.go:145 +0xcb
    github.com/ceph/ceph-csi/e2e.createPVCAndAppBinding(0xc000048460, 0x1b, 0xc000048480, 0x1b, 0xc000385e40, 0xa, 0x2, 0xc000554c30, 0x4fb151, 0x317cb00)
    	/go/src/github.com/ceph/ceph-csi/e2e/utils.go:186 +0x108
    github.com/ceph/ceph-csi/e2e.validatePVCAndAppBinding(0xc000048460, 0x1b, 0xc000048480, 0x1b, 0xc000385e40, 0xa000, 0x3e)
    	/go/src/github.com/ceph/ceph-csi/e2e/utils.go:195 +0x66
    github.com/ceph/ceph-csi/e2e.glob..func1.3.1.3()
    	/go/src/github.com/ceph/ceph-csi/e2e/cephfs.go:286 +0x52
    github.com/ceph/ceph-csi/e2e.glob..func1.3.1()
    	/go/src/github.com/ceph/ceph-csi/e2e/cephfs.go:285 +0x15e5
    github.com/ceph/ceph-csi/e2e.TestE2E(0xc000103680)
    	/go/src/github.com/ceph/ceph-csi/e2e/e2e_test.go:50 +0xa9
    testing.tRunner(0xc000103680, 0x205e2a0)
    	/usr/local/go/src/testing/testing.go:1193 +0xef
    created by testing.(*T).Run
    	/usr/local/go/src/testing/testing.go:1238 +0x2b3
```

Rakshith-R commented on e2e/pvc.go:
```diff
@@ -61,14 +61,14 @@ func createPVCAndvalidatePV(c kubernetes.Interface, pvc *v1.PersistentVolumeClai
 
 		pv, err = c.CoreV1().PersistentVolumes().Get(context.TODO(), pvc.Spec.VolumeName, metav1.GetOptions{})
 		if err != nil {
+			if isRetryableAPIError(err) {
```
[e2e run](https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e-helm_k8s-1.19/detail/mini-e2e-helm_k8s-1.19/1999/pipeline/) ([logs](https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/rest/organizations/jenkins/pipelines/mini-e2e-helm_k8s-1.19/runs/1999/nodes/98/log/?start=0))
```
Jul 13 18:21:12.215: INFO: Error getting pvc in namespace: '': etcdserver: request timed out
Jul 13 18:21:12.981: INFO: waiting for PVC  (102 seconds elapsed)
Jul 13 18:21:12.981: INFO: Error getting pvc in namespace: '': an empty namespace may not be set when a resource name is provided
Jul 13 18:21:12.981: INFO: failed to create PVC (rbd-32740): failed to get pvc: an empty namespace may not be set when a resource name is provided
```
@nixpanic Looks like we are overwriting on existing `pvc` variable with a empty `&PVC{}` if there's a error during `Get` at line 46, probably need to use another variable to avoid this.

nixpanic replied:
Indeed! That is really ugly :disappointed:

nixpanic replied:
now using an extra `namespace` variable to store the value that is initially used (just like `name` refers to `pvc.Name`).

nixpanic committed:
e2e: prevent re-using empty pvc object

When an error occurs, the pvc object is overwritten in the
PollImmediate() loop. Re-using the pvc.Namespace results in error
messages like

    Error getting pvc in namespace: '': an empty namespace may not be set when a resource name is provided

and prevents the retry by PollImmediate() to never succeed. Storing the
namespace in a local variable prevents this from happening.

Reported-by: Rakshith R <rar@redhat.com>
Signed-off-by: Niels de Vos <ndevos@redhat.com>
